process/process.go

```diff
@@ -0,0 +1,380 @@
+	go func() {
+		time.Sleep(duration)
+
+		p.mu.Lock()
+		defer p.mu.Unlock()
+
+		p.timeoutPrecommit(height, round)
+	}()
+}
+
+func (p *Process) checkProposeInCurrentHeightAndRoundWithPrevotes() {
+	// upon Propose{currentHeight, currentRound, block, validRound} from Schedule(currentHeight, currentRound)
+	m := p.state.Proposals.QueryByHeightRoundSignatory(p.state.CurrentHeight, p.state.CurrentRound, p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound))
+	if m == nil {
+		return
+	}
+	propose := m.(*Propose)
+
+	if propose.ValidRound() > block.InvalidRound {
+		// and 2f+1 Prevote{currentHeight, validRound, blockHash}
+		n := p.state.Prevotes.QueryByHeightRoundBlockHash(p.state.CurrentHeight, propose.ValidRound(), propose.BlockHash())
+		if n > 2*p.state.Prevotes.F() {
+			// while step = StepPropose and validRound >= 0 and validRound < currentRound
+			if p.state.CurrentStep == StepPropose && propose.ValidRound() < p.state.CurrentRound {
+				if p.validator.IsBlockValid(propose.Block()) && (p.state.LockedRound <= propose.ValidRound() || p.state.LockedBlock.Equal(propose.Block())) {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						propose.Block().Hash(),
+					))
+				} else {
+					p.broadcaster.Broadcast(NewPrevote(
+						p.state.CurrentHeight,
+						p.state.CurrentRound,
+						block.InvalidHash,
+					))
+				}
+				p.state.CurrentStep = StepPrevote
+			}
+		}
+	}
+}
+
+// checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime must only be
+// called when a Propose and 2f+1 Prevotes has been seen for the first time at
+// the current `block.Height` and `block.Round`. This can happen when a Propose
+// is seen for the first time at the current `block.Height` and `block.Round`,
+// or, when a Prevote is seen for the first time at the current `block.Height`
+// and `block.Round`.
+func (p *Process) checkProposeInCurrentHeightAndRoundWithPrevotesForTheFirstTime() {
+	// upon Propose{currentHeight, currentRound, block, *} from Schedule(currentHeight, currentRound)
+	m := p.state.Proposals.QueryByHeightRoundSignatory(p.state.CurrentHeight, p.state.CurrentRound, p.scheduler.Schedule(p.state.CurrentHeight, p.state.CurrentRound))
+	if m == nil {
+		return
+	}
+	propose := m.(*Propose)
+
+	// and 2f+1 Prevote{currentHeight, currentRound, blockHash} while Validate(block) and step >= StepPrevote for the first time
+	n := p.state.Prevotes.QueryByHeightRoundBlockHash(p.state.CurrentHeight, p.state.CurrentRound, propose.BlockHash())
+	if n > 2*p.state.Prevotes.F() {
+		if p.state.CurrentStep >= StepPrevote && p.validator.IsBlockValid(propose.Block()) {
+			if p.state.CurrentStep == StepPrevote {
+				p.state.LockedBlock = propose.Block()
+				p.state.LockedRound = p.state.CurrentRound
+				p.broadcaster.Broadcast(NewPrecommit(
+					p.state.CurrentHeight,
+					p.state.CurrentRound,
+					propose.Block().Hash(),
+				))
```

Comment:
I think this step was missed here:
41: step<sub>p</sub>←precommit

Reply:
Good catch. Fixed.